Modules/_zoneinfo.c

```diff
@@ -2500,28 +2623,15 @@ new_weak_cache(void)
 }
 
 static int
-initialize_caches(void)
+initialize_caches(zoneinfo_state *state)
 {
-    // TODO: Move to a PyModule_GetState / PEP 573 based caching system.
-    if (TIMEDELTA_CACHE == NULL) {
-        TIMEDELTA_CACHE = PyDict_New();
-    }
-    else {
-        Py_INCREF(TIMEDELTA_CACHE);
-    }
-
-    if (TIMEDELTA_CACHE == NULL) {
+    state->TIMEDELTA_CACHE = PyDict_New();
```

Comment:
`Py_mod_exec` is called to initialise new module objects only, so we're passed an unitialised module state. We don't need to check for reinitialisation.

Reply:
Is something like this sufficient, of were you thinking of adding a `cache_behaviour` section to the docs as well?

```diff
diff --git a/Doc/library/zoneinfo.rst b/Doc/library/zoneinfo.rst
index d2e5619e7e..f66208568a 100644
--- a/Doc/library/zoneinfo.rst
+++ b/Doc/library/zoneinfo.rst
@@ -238,13 +238,6 @@ The following class methods are also available:
 
     .. TODO: Add "See `cache_behavior`_" reference when that section is ready.
 
-    .. warning::
-
-        Invoking this function may change the semantics of datetimes using
-        ``ZoneInfo`` in surprising ways; this modifies process-wide global state
-        and thus may have wide-ranging effects. Only use it if you know that you
-        need to.
-
 The class has one attribute:
 
 .. attribute:: ZoneInfo.key
```

Reply:
I just meant that we should have a comment in the function like `// This function is not idempotent and must be callled on a new module object`

Reply:
Sorry, I misunderstood you. Thanks for the enlightening comment (https://github.com/python/cpython/pull/99218#discussion_r1107382566).

Reply:
Sure, I can add such a comment, but isn't it obvious since it's called from the module exec slot?

Reply:
> Sure, I can add such a comment, but isn't it obvious since it's called from the module exec slot?

Maybe if you are intimately familiar with the how the module exec machinery works, but I write C extensions so rarely that every time **I** am mucking around in the code, I've completely forgotten how everything works and what the guarantees and assumptions are (if I ever knew them in the first place), and I think that my experience is pretty common — or maybe the most unusual part of it is that I write C extensions *at all*.

Additionally, while unlikely, it's not impossible to imagine some future refactoring where someone changes the cache clearing mechanism to null out some of these module fields, then says, "Oh whenever we need the caches warmed we'll just call `initialize_caches`", or breaks up the function into smaller reusable parts, forgetting that it was originally intended to be called *exactly* one time per module.

Anyway, it's a bit belt-and-suspenders I know, but I think little signposts about one's assumptions are useful to future developers (including future us).